### ISSUE TITLE
Proposal: improved sd-boot interation with Windows+BitLocker

### ISSUE THREAD

**Issue by danderson** [systemd/systemd#21891]
I'm sure the background is well known, but I'll repeat it anyway, just in case: Windows with Bitlocker enabled is hostile to any changes to the Secure Boot chain of trust, because the bitlocker key is sealed to TPM PCRs, including the bootloader chain used to start Windows. IOW, Bitlocker expects that Windows is the only one trying to modify the bootloader config, and freaks out if it changes unexpectedly.

The canonical "fix" for this recommended by distros is to disable Bitlocker during linux install, install sd-boot or some other bootloader, then reenable Bitlocker in Windows. Bitlocker will then seal the decryption key against PCR state that represents "Windows chainloaded by sd-boot/grub/whatever", and it will be happy... Until the next bootloader code update, where it will freak out again.

This is especially problematic now with systemd 250, because of the new default-enabled unit that upgrades sd-boot when necessary. That means that a dual-booting system with Bitlocker enabled will randomly panic and lock users out of Windows without an obvious cause. Of course, distros can turn that functionality off, but it's very nice to have and it would be better to leave it enabled.

To fix this more permanently, I propose to change how sd-boot's automatic entry for windows works: instead of loading the Windows loader EFI binary and executing it, sd-boot should set the NextBoot UEFI variable to the Windows loader, and reset the system. 

Because of the reset, the PCR state will be cleared, and with the NextBoot variable set, UEFI will load straight into the Windows loader. That way, from Bitlocker's POV, there are no surprises in the Secure Boot state, it's only UEFI + various Windows software. And then on the boot after that, of course, NextBoot will be cleared and the system will come back to sd-boot's menu.

The main downside is that booting into Windows will take longer, because it requires a warm reset and walking through all the UEFI code again. However, on most desktops and laptops, UEFI execution is pretty quick if the management engine/security processor is already running (which is the case for warm boots). It would suck more for dual-booting servers that have very slow firmware init, but IMO dual-booting server hardware is very rare. Booting is also pretty rare these days compared to sleep, so a small extra delay to get reliable Windows booting seems like a good tradeoff to me.

Alternatives considered:
 - Keep the current behavior, where Bitlocker freaks out at random times. This is a poor user experience, and I worry that they will blame linux/systemd for the problem.
 - Get Microsoft to offer APIs that let linux cooperate better with Bitlocker. Seems pretty unlikely to me :(
 - Tell everyone to edit the windows registry and remove the bootloader PCRs from the Bitlocker integrity list (there's a registry key that lists which PCRs it should use). This would work reliably, but nerfs Windows's boot security by allowing hostile code to access the drive encryption key.

WDYT?

**Comment by medhefgo** [comment]
According to the group policy description, this should already work under secure boot as it will only use PCR 0, 2 and 7. sd-boot would have to be signed by the same keys across updates that are known to the firmware or shim (if used, hash based enrolling shouldn't work for this). Sadly, I lack a TPM to test this properly.

Are you using secure boot?

Of course, this still is of use when secure boot is off. I already have some working code, but I'd rather only enable it in situations where this is known to be needed, and secure boot enabled should make this unnecessary.

**Comment by danderson** [comment]
Oh, interesting, I was going off documentation that predates Secure Boot's use of PCR 7. And indeed, in the case where Secure Boot is enabled, if Bitlocker only uses 0,2,7 to seal the key, that should be fine.

I do have a couple machines with TPMs, although ironically I think I had to disable secure boot on all of them because the distros I run haven't figured out boot path security yet. So, I guess that's why I'm hitting the painful Bitlocker experience. But I can wipe one of my laptops and try to upgrade sd-boot with bitlocker and secure boot enabled, to confirm that it works correctly.

**Comment by medhefgo** [comment]
> But I can wipe one of my laptops and try to upgrade sd-boot with bitlocker and secure boot enabled, to confirm that it works correctly.

No need for that. I'll try a VM first. Though, real system confirmation always welcome.

**Comment by medhefgo** [comment]
I tested this under QEMU with custom secure boot keys. The default is 0,2,4,11 if *Secure Boot State (PCR7)* support is not available, otherwise it's 7,11 (according to the group policy description). The decision to use PCR7 seems to be either broken or PCR7 isn't supported by swtpm/OVMF/libvirt (I doubt that) so it's not chosen automatically.

If I manually set the policy to 0,2,7,11 I can successfully update the boot loader without needed the recovery key. 

I would love to see the default PCRs selected on real (and recent) hardware with TPM and secure boot enabled (`manage-bde -protectors C: -get`)…

So I guess for now, it should be activated in all BitLocker+TPM cases.
